Mathlib/Order/KrullDimension.lean

```diff
@@ -30,40 +40,104 @@ in this file would generalize as well. But we don't think it would be useful, so
 
 /--
 The **Krull dimension** of a preorder `α` is the supremum of the rightmost index of all relation
-series of `α` order by `<`. If there is no series `a₀ < a₁ < ... < aₙ` in `α`, then its Krull
+series of `α` ordered by `<`. If there is no series `a₀ < a₁ < ... < aₙ` in `α`, then its Krull
 dimension is defined to be negative infinity; if the length of all series `a₀ < a₁ < ... < aₙ` is
 unbounded, its Krull dimension is defined to be positive infinity.
 -/
-noncomputable def krullDim : WithBot (WithTop ℕ) :=
+noncomputable def krullDim (α : Type*) [Preorder α] : WithBot ℕ∞ :=
   ⨆ (p : LTSeries α), p.length
 
 /--
-Height of an element `a` of a preordered set `α` is the Krull dimension of the subset `(-∞, a]`
+The **height** of an element `a` in a preorder `α` is the supremum of the rightmost index of all
+relation series of `α` ordered by `<` and ending below or at `a`.
 -/
-noncomputable def height (a : α) : WithBot (WithTop ℕ) := krullDim (Set.Iic a)
+noncomputable def height {α : Type*} [Preorder α] (a : α) : ℕ∞ :=
+  ⨆ (p : LTSeries α) (_ : p.last ≤ a), p.length
 
 /--
-Coheight of an element `a` of a pre-ordered set `α` is the Krull dimension of the subset `[a, +∞)`
+The **coheight** of an element `a` in a preorder `α` is the supremum of the rightmost index of all
+relation series of `α` ordered by `<` and beginning with `a`.
+
+The definition of `coheight` is via the `height` in the dual order, in order to easily transfer
+theorems between `height` and `coheight`.
 -/
-noncomputable def coheight (a : α) : WithBot (WithTop ℕ) := krullDim (Set.Ici a)
+noncomputable def coheight {α : Type*} [Preorder α] (a : α) : ℕ∞ := height (α := αᵒᵈ) a
 
 end definitions
 
+section height
+
+variable {α β : Type*}
+
+variable [Preorder α] [Preorder β]
+
+/-!
+## Height
+-/
+
+lemma height_le_iff (x : α) (n : ℕ∞) :
+    height x ≤ n ↔ ∀ (p : LTSeries α), p.last ≤ x → p.length ≤ n := by
+  simp [height, iSup_le_iff]
+
+lemma height_le (x : α) (n : ℕ∞) :
+    (∀ (p : LTSeries α), p.last ≤ x → p.length ≤ n) → height x ≤ n :=
+  (height_le_iff x n).mpr
+
+lemma length_le_height (x : α) (p : LTSeries α) (hlast : p.last ≤ x) :
+    p.length ≤ height x := by
+  by_cases hlen0 : p.length ≠ 0
+  · let p' := p.eraseLast.snoc x (by
+      apply lt_of_lt_of_le
+      · apply p.step ⟨p.length - 1, by omega⟩
+      · convert hlast
+        simp only [Fin.succ_mk, Nat.succ_eq_add_one, RelSeries.last, Fin.last]
+        congr; omega)
+    suffices p'.length ≤ height x by
+      simp [p'] at this
+      convert this
+      norm_cast
+      omega
+    refine le_iSup₂_of_le p' ?_ le_rfl
+    simp [p']
+  · simp_all
+
+lemma length_le_height_last (p : LTSeries α) : p.length ≤ height p.last :=
+  length_le_height _ p le_rfl
+
+lemma height_mono : Monotone (α := α) height :=
+  fun _ _ hab ↦ biSup_mono (fun _ hla => hla.trans hab)
+
+@[gcongr] protected lemma _root_.GCongr.height_le_height (a b : α) (hab : a ≤ b) :
+    height a ≤ height b := height_mono hab
+
+end height
```

Comment:
I would expect the `coheight` API to come here, just after the `height` one

Reply:
The coheight API is in a separate PR. Wouldn’t it be easier to review all PRs about `height` API first (so they are smaller and there is less noise), and then one PR that does the busy work of adding the coheight?

Reply:
Ah yeah sure. I guess I'm just a bit confused as to why you are also changing the definition of `height` but not providing its API

Reply:
Fair point. I guess because the old definition is even more strange, as it has the “wrong” type, so it’s still a step in the right direction.